gh-121342: Fixed `pkgutil.iter_zipimport_modules` on an invalidated cache (#121342)

beachmachine opened the pull request:
It is no longer safe to directly access `zipimport._zip_directory_cache` since #103208. It is not guaranteed that the cache is acutally filled. This changed fixes this by using the internal method `_get_files` instead, which may not be the best solution, but fixes the issue.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-121342 -->
* Issue: gh-121342
<!-- /gh-issue-number -->

beachmachine committed:
gh-121342: Fixed `pkgutil.iter_zipimport_modules` on an invalidated cache (#121342)

It is no longer safe to directly access `zipimport._zip_directory_cache` since #103208. It is
not guaranteed that the cache is acutally filled. This changed fixes this by using the
internal method `_get_files` instead, which may not be the best solution, but fixes the issue.

ZeroIntensity commented on Lib/test/test_pkgutil.py:
```diff
@@ -522,6 +522,54 @@ def test_mixed_namespace(self):
         del sys.modules['foo.bar']
         del sys.modules['foo.baz']
 
+
```
Only one blank line between methods, per [PEP 8](https://peps.python.org/pep-0008/)
```suggestion
```

beachmachine replied:
Removed the blank line. Also, I've learned that you prefer not to have force-pushes on CPython PRs. I'll respect that for future PRs. :)

ZeroIntensity commented on Lib/test/test_pkgutil.py:
```diff
@@ -522,6 +522,54 @@ def test_mixed_namespace(self):
+
+        # Cleanup
+        os.remove(tmp_file_zip.filename)
+
```
Same here.
```suggestion
```

beachmachine replied:
Removed the blank line.

ZeroIntensity commented on Lib/test/test_pkgutil.py:
```diff
@@ -522,6 +522,54 @@ def test_mixed_namespace(self):
+
+        # Cleanup
+        os.remove(tmp_file_zip.filename)
+
```
```suggestion
```

beachmachine replied:
Removed the blank line.

ZeroIntensity reviewed (changes requested):
Only a few formatting nitpicks. Failing CI is caused by `pkgutil.iter_zipimport_modules` being undocumented. Ideally, you should document it with this PR.

beachmachine committed:
gh-121342: Unreferenced `pkgutil.iter_zipimport_modules` in NEWS.d

As `pkgutil.iter_zipimport_modules` isn't supposed to be consumed directly
by anything other than `pkgutils` itself, and is registered via as an
importer module for `zipimporter`, I've decided to not include it in the
public documentation. Thus, the `:func:` reference
of `pkgutil.iter_zipimport_modules` in the news file is removed now.

beachmachine committed:
gh-121342: Make the Sphinx linter happy again
